runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/testing/TestDataflowRunner.java

```diff
@@ -126,6 +126,10 @@ DataflowPipelineJob run(Pipeline pipeline, DataflowRunner runner) {
               for (;;) {
                 JobMetrics metrics = getJobMetrics(job);
                 Optional<Boolean> success = checkForPAssertSuccess(job, metrics);
+                if (messageHandler.hasSeenError()) {
```

Comment:
Just tacking this on: perhaps we also shouldn't check for PAssert success here, but only max watermark. If the PAsserts all succeed before an error is witnessed, the test should still fail.

Downside is losing tests that are designed to never terminate, but on the other hand, I'm not sure we have those.